framework/include/actions/ActionWarehouse.h

```diff
@@ -101,6 +101,64 @@ class ActionWarehouse :
   const std::vector<Action *> & getActionsByName(const std::string & task) const;
 
   /**
+   * Retrieve an action with its name and the desired type.
+   * @param name The action name.
+   */
+  template <class T>
+  const T & getAction(const std::string & name)
+    {
+      T* p = NULL;
+      for (std::vector<std::string>::iterator it = _ordered_names.begin(); it != _ordered_names.end(); ++it)
+      {
+        std::string task = *it;
+        for (ActionIterator act_iter = actionBlocksWithActionBegin(task);
+             act_iter != actionBlocksWithActionEnd(task);
+             ++act_iter)
+        {
+          if ((*act_iter)->name() == name)
+          {
+            p = dynamic_cast<T*>(*act_iter);
+            if (!p)
+              mooseError("The action with name being "+name+" is not in the desired type");
+            else
+              break;
+          }
+        }
+      }
+      if (!p)
+        mooseError("Action with name being "+name+" does not exist");
+      return *p;
+    }
+
+  /**
+   * Retrieve all actions in a specific type.
+   */
+  template <class T>
+  std::vector<const T *> getActions()
```

Comment:
We don't grab any other type (besides UserObjects) through templated arguments. The base class supports the interface (getting things by name) so why change this here? It just doesn't feel very MOOSE-y. Yeah I'm sure it's convenient for you to get a list of Actions by the specified type but you still have to loop over them all and dynamic cast them all. I'd rather see that kind of stuff in user-code (not MOOSE).